src/main/java/me/confor/velocity/chat/modules/GlobalChat.java

```diff
@@ -66,57 +64,61 @@ public void onPlayerChat(PlayerChatEvent event) {
-            sendMessage(msg,currentServer.get().getServer());
-        else
             sendMessage(msg);
+        else
+            sendMessage(msg, currentServer.get().getServer());
 
         if (config.GLOBAL_CHAT_TO_CONSOLE)
             this.logger.info("GLOBAL: <{}> {}", player, message);
 
-        if (!config.GLOBAL_CHAT_PASSTHROUGH)
+        /*
+        broken >=1.19.1
+
+        if (config.GLOBAL_CHAT_PASSTHROUGH)
             event.setResult(ChatResult.denied());
+         */
     }
 
     @Subscribe
-    public void onConnect(LoginEvent event) {
-        if (!config.JOIN_ENABLE)
-            return;
+    public void onConnect(ServerConnectedEvent event) {
+        RegisteredServer currentServer = event.getServer();
+        Optional<RegisteredServer> previousServer = event.getPreviousServer();
 
         String player = event.getPlayer().getUsername();
+        String server = currentServer.getServerInfo().getName();
 
-        Component msg = parseMessage(config.JOIN_FORMAT, List.of(
-                new ChatTemplate("player", player, false)
-        ));
-
-        sendMessage(msg);
-    }
-
-    @Subscribe
-    public void onDisconnect(DisconnectEvent event) {
-        if (!config.QUIT_ENABLE)
-            return;
+        Component msg;
 
-        Optional<ServerConnection> currentServer = event.getPlayer().getCurrentServer();
-        if (currentServer.isEmpty())
-            return;
+        if (previousServer.isPresent()) {
+            if (!config.SWITCH_ENABLE)
+                return;
 
-        String player = event.getPlayer().getUsername();
-        String server = currentServer.get().getServerInfo().getName();
+            msg = parseMessage(config.SWITCH_FORMAT, List.of(
+                new ChatTemplate("player", player, false),
+                new ChatTemplate("server", server, false),
+                new ChatTemplate("previous_server", previousServer.get().getServerInfo().getName(), false)
+            ));
+        } else {
+            if (!config.JOIN_ENABLE)
+                return;
 
-        Component msg = parseMessage(config.QUIT_FORMAT, List.of(
+            msg = parseMessage(config.JOIN_FORMAT, List.of(
                 new ChatTemplate("player", player, false),
                 new ChatTemplate("server", server, false)
-        ));
+            ));
 
-        if (config.GLOBAL_CHAT_PASSTHROUGH)
-            sendMessage(msg,currentServer.get().getServer());
-        else
-            sendMessage(msg);
+            if (!config.GLOBAL_CHAT_PASSTHROUGH) {
+                sendMessage(msg, currentServer);
+                return;
+            }
+        }
+
+        sendMessage(msg);
     }
 
     @Subscribe
-    public void onServerConnect(ServerPostConnectEvent event) {
-        if (!config.SWITCH_ENABLE)
+    public void onDisconnect(DisconnectEvent event) {
+        if (!config.QUIT_ENABLE)
             return;
 
         Optional<ServerConnection> currentServer = event.getPlayer().getCurrentServer();
```

Comment:
> I think it may be possible for the player to disconnect without the current server being known.

Yes, I think that would happen when a backend server is unreachable. A similar plugin I've worked on, [VelocityDiscord](https://github.com/fooooooooooooooo/VelocityDiscord), implemented something to handle that as well: https://github.com/fooooooooooooooo/VelocityDiscord/commit/64ef162e6afa9e051646c0faa6dceae1f1810ebc

Reply:
In the past I could reproduce this when connecting through velocity to a server with the wrong client version (minecraft 1.17 to a 1.18 server). Strange things happen to events when you immediately fail to connect.